### ISSUE TITLE
Terraform remote state S3 does not honour role_arn set via aws profile

### ISSUE THREAD

**Issue by pieterza** [hashicorp/terraform#26995]
### Terraform Version
Terraform v0.13.5
Terraform v0.12.29


#Nutshell:

The backend has an option called 'profile'
This should map to a profile configured inside the aws config (~/.aws/credentials)
AWS config profiles can also contain role_arn's to assume
The terraform backend does not assume the profile's role_arn

Proven by manually adding the arn to the backend config, which then works.

### Terraform Configuration Files

```
terraform {                                                                                                                                                                                                        
  backend "s3" {                                                                                                                                                                                                     
    bucket   = "foo-tf-state"                                                                                                                                                                                       
    key     = "foo-bar-af1-web-app/terraform.tfstate"                                                                                                                                                              
    region  = "eu-west-1"                                                                                                                                                                                            
    encrypt    = true                                                                                                                                                                                                
    kms_key_id = "alias/base/s3"                                                                                                                                                                                     
    dynamodb_table = "foo-tf-state"                                                                                                                                                                                 
    profile = "foo"                                                                                                                                                                                                                                                                                                                                                                           
  }                                                                                                                                                                                                              }                
```

### AWS config

```
[foo]
region = eu-west-1
credential_source = Ec2InstanceMetadata
role_arn = arn:aws:iam::1234REDACT:role/foo-cross-venture-role-tf-state
```


### Debug Output
`terraform init`

```
Initializing modules...

Initializing the backend...
Error refreshing state: AccessDenied: Access Denied
        status code: 403, request id: E7F3652D3408C24E, host id: g0HJJ1OlT+HCpk0AuRnxTnqEI+fI03xdTREDACTJByb0MB1fDTe8moSIaVbMuceIelsjqinU=
bash-5.0#
```

### Expected Behavior

Init should succeed by assuming the role specified in ~/aws/credentials: [foo] profile.


### Actual Behavior

Terraform init only uses the instance profile and does not also assume the role.

### Steps to Reproduce

Have AWS credentials like:

```
[foo]
region = eu-west-1
credential_source = Ec2InstanceMetadata
role_arn = arn:aws:iam::1234REDACT:role/foo-cross-venture-role-tf-state
```

Have backend.tf like:

```
terraform {                                                                                                                                                                                                        
  backend "s3" {                                                                                                                                                                                                     
    bucket   = "foo-tf-state"                                                                                                                                                                                       
    key     = "foo-bar-af1-web-app/terraform.tfstate"                                                                                                                                                              
    region  = "eu-west-1"                                                                                                                                                                                            
    encrypt    = true                                                                                                                                                                                                
    kms_key_id = "alias/base/s3"                                                                                                                                                                                     
    dynamodb_table = "foo-tf-state"                                                                                                                                                                                 
    profile = "foo"                                                                                                                                                                                                                                                                                                                                                                           
  }                                                                                                                                                                                                              }            
```

Run terraform init.

What happens is that terraform init does not honor the assume_role set via aws config, as suggested by accept a profile option.

We can prove this by simply adding:

role_arn     = "arn:aws:iam::1234REDACT:role/foo-cross-venture-role-tf-state" to the backend.tf like so:

```
terraform {                                                                                                                                                                                                        
  backend "s3" {                                                                                                                                                                                                     
    bucket   = "foo-tf-state"                                                                                                                                                                                       
    key     = "foo-bar-af1-web-app/terraform.tfstate"                                                                                                                                                              
    region  = "eu-west-1"                                                                                                                                                                                            
    encrypt    = true                                                                                                                                                                                                
    kms_key_id = "alias/base/s3"                                                                                                                                                                                     
    dynamodb_table = "foo-tf-state"                                                                                                                                                                                 
    profile = "foo"           
    role_arn     = "arn:aws:iam::1234REDACT:role/foo-cross-venture-role-tf-state                                                                                                                                                                                                                                                                                                                                                
  }                                                                                                                                                                                                              }
```

The result is that the init works.

### Additional Context

The terraform s3 backend accepts a 'profile' option which maps to an AWS profile.
AWS profile has an role_arn key, which is ignored by terraform
Proven by also duplicating the role_arn inside the backend.tf file.

The backend should assume whatever roles are set in the ~/aws/credentials file matching the profile specified.

**Comment by pieterza** [comment]
> I'm facing exactly the same issue, using Terraform from an ec2 instance with an instance profile attached.
> 
> It seems that Terraform doesn't first assume **the role specified in the config file** and then the one specified in the s3 backend, but that tries directly to assume the role specified in the backend, and due to how my IAM policies are configured fails.
> 
> My config file located in .aws directory:
> 
> ```
> [profile example]
> credential_source = Ec2InstanceMetadata
> role_arn = arn:aws:iam::account_id:role/first_role_to_assume
> region = eu-west-1
> ```
> 
> My s3 backend configuration:
> 
> ```
> backend "s3" {
>     bucket         = "example"
>     key            = "example.tfstate"
>     region         = "eu-west-1"
>     dynamodb_table = "example"
>     profile        = "example"
>     role_arn       = "arn:aws:iam::account_id:role/second_role_to_assume_after_having_assumed_the_first_one"
> }
> ```
> 
> @pieterza have you found any workaround?

Yeah, but it's hacky.

Instead of relying on the ~/.aws/credentials' role_arn (which should be assumed via backend's profile setting, but is not), we just set role_arn inside terraform, multiple times and in our _remotes.tf
I believe I also added a bash wrapper to dynamically transform some files whenever our terraform wrapper script runs.

**Comment by pieterza** [comment]
Any update on this one?

**Comment by thiagolsfortunato** [comment]
Any news?

**Comment by amcsi** [comment]
Also sadly you can't use `role_arn` dynamically to allow for different devs/envs to use different roles:

```
attila-Predator-Kubuntu :: src/Upstack/infrastructure ‹upstack-application-profile*› % terraform init -reconfigure                                                                   1 ↵
Initializing modules...

Initializing the backend...
╷
│ Error: Variables not allowed
│ 
│   on main.tf line 16, in terraform:
│   16:     role_arn = var.app_infra_role_arn
│ 
│ Variables may not be used here.
╵
```

So please fix this issue, otherwise I have to pick between either having security (assuming roles for provisioning rather than using an iam user directly), or being able to have dynamic auth creds; I can't have both at the same time.

**Comment by thenaturalist** [comment]
Hi there, it's been well over 1.5 years. Is there an ETA on this?

**Comment by c240amg** [comment]
Yup this is a bad blocker for enterprise users - we're using Control Tower to deploy multiple accounts and having this issue fixed would make life so much easier.....

**Comment by crw** [comment]
Hi @c240amg, if you are a Terraform Enterprise user, you can also contact your account team or use https://support.hashicorp.com/hc/en-us/requests/new to raise this issue with your account team.

**Comment by thameezb** [comment]
Any movement on this?

**Comment by Tyler-Rendina** [comment]
Can't believe this is an issue, here's my redacted logs and related code up to the incorrect cred request to aws:
```log
AWS_PROFILE=prod TF_LOG=DEBUG terraform init
2023-08-02T17:56:10.675-0400 [INFO]  Terraform version: 1.5.4
2023-08-02T17:56:10.675-0400 [DEBUG] using github.com/hashicorp/go-tfe v1.26.0
2023-08-02T17:56:10.675-0400 [DEBUG] using github.com/hashicorp/hcl/v2 v2.16.2
2023-08-02T17:56:10.675-0400 [DEBUG] using github.com/hashicorp/terraform-svchost v0.1.0
2023-08-02T17:56:10.675-0400 [DEBUG] using github.com/zclconf/go-cty v1.12.2
2023-08-02T17:56:10.675-0400 [INFO]  Go runtime version: go1.20.6
2023-08-02T17:56:10.675-0400 [INFO]  CLI args: []string{"terraform", "init"}
2023-08-02T17:56:10.675-0400 [DEBUG] Attempting to open CLI config file: /Users//.terraformrc
2023-08-02T17:56:10.675-0400 [DEBUG] File doesn't exist, but doesn't need to. Ignoring.
2023-08-02T17:56:10.675-0400 [DEBUG] ignoring non-existing provider search directory terraform.d/plugins
2023-08-02T17:56:10.675-0400 [DEBUG] ignoring non-existing provider search directory /Users//.terraform.d/plugins
2023-08-02T17:56:10.675-0400 [DEBUG] ignoring non-existing provider search directory /Users//Library/Application Support/io.terraform/plugins
2023-08-02T17:56:10.675-0400 [DEBUG] ignoring non-existing provider search directory /Library/Application Support/io.terraform/plugins
2023-08-02T17:56:10.676-0400 [INFO]  CLI command args: []string{"init"}

Initializing the backend...
2023-08-02T17:56:10.687-0400 [INFO]  AWS Auth provider used: "EnvProvider"
2023-08-02T17:56:10.688-0400 [DEBUG] Trying to get account information via sts:GetCallerIdentity
2023-08-02T17:56:10.690-0400 [DEBUG] [aws-sdk-go] DEBUG: Request sts/GetCallerIdentity Details:
```

main.tf
```hcl
terraform {
  backend "s3" {
    profile = "prod"
    bucket = 
    key    = 
    region = "us-east-2"
    dynamodb_table = 
    encrypt = true
  }
}
provider "aws" {
  profile = "prod"
  default_tags {
    tags = var.tags
  }
}
```

**Comment by gdavison** [comment]
This will be addressed in Terraform v1.6. If you still see this error when v1.6 is released (or you try any of the v1.6 beta releases), please open a new issue
